ring/token_range_test.go

```diff
@@ -0,0 +1,306 @@
+				}
+			}
+		})
+	}
+}
+
+func BenchmarkGetTokenRangesForInstance(b *testing.B) {
+	instancesPerZone := []int{1, 3, 9, 27, 81, 243, 729}
+
+	for _, n := range instancesPerZone {
+		b.Run(fmt.Sprintf("%d_instancesperzone", n), func(b *testing.B) {
+			benchmarkGetTokenRangesForInstance(b, n)
+		})
+	}
+}
+
+func benchmarkGetTokenRangesForInstance(b *testing.B, instancesPerZone int) {
+	numZones := 3
+	numTokens := 512
+
+	b.ResetTimer()
+
+	for n := 0; n < b.N; n++ {
+		b.StopTimer()
+		// Initialise the ring.
+		ringDesc := &Desc{Ingesters: generateRingInstances(instancesPerZone*numZones, numZones, numTokens)}
+		ring := Ring{
+			cfg:                  Config{HeartbeatTimeout: time.Hour, ZoneAwarenessEnabled: true, SubringCacheDisabled: true, ReplicationFactor: numZones},
+			ringDesc:             ringDesc,
+			ringTokens:           ringDesc.GetTokens(),
+			ringTokensByZone:     ringDesc.getTokensByZone(),
+			ringInstanceByToken:  ringDesc.getTokensInfo(),
+			ringZones:            getZones(ringDesc.getTokensByZone()),
+			shuffledSubringCache: map[subringCacheKey]*Ring{},
+			strategy:             NewDefaultReplicationStrategy(),
+			lastTopologyChange:   time.Now(),
+		}
+		b.StartTimer()
+
+		_, _ = ring.GetTokenRangesForInstance("instance-1")
+	}
+}
+
+func TestCheckingOfKeyOwnership(t *testing.T) {
+	const instancesPerZone = 100
+	const numZones = 3
+	const numTokens = 512
+	const replicationFactor = numZones // This is the only config supported by GetTokenRangesForInstance right now.
+
+	// Generate users with different number of tokens
+	userTokens := map[string][]uint32{}
+	shardSizes := map[string]int{}
+	for _, cnt := range []int{1000, 5000, 10000, 25000, 50000, 100000, 250000, 500000} {
+		uid := fmt.Sprintf("%dk", cnt/1000)
+		userTokens[uid] = GenerateTokens(cnt, nil)
+
+		shardSize := cnt / 7500
+		shardSize = (shardSize / numZones) * numZones // round down to numZones
+		if shardSize < numZones {
+			shardSize = numZones
+		}
+		shardSizes[uid] = shardSize
+	}
+
+	// Generate ring
+	ringDesc := &Desc{Ingesters: generateRingInstances(instancesPerZone*numZones, numZones, numTokens)}
+	ring := Ring{
+		cfg:                  Config{HeartbeatTimeout: time.Hour, ZoneAwarenessEnabled: true, SubringCacheDisabled: false, ReplicationFactor: replicationFactor},
+		ringDesc:             ringDesc,
+		ringTokens:           ringDesc.GetTokens(),
+		ringTokensByZone:     ringDesc.getTokensByZone(),
+		ringInstanceByToken:  ringDesc.getTokensInfo(),
+		ringZones:            getZones(ringDesc.getTokensByZone()),
+		shuffledSubringCache: map[subringCacheKey]*Ring{},
+		strategy:             NewDefaultReplicationStrategy(),
+		lastTopologyChange:   time.Now(),
+	}
+
+	for uid, tokens := range userTokens {
+		shardSize := shardSizes[uid]
+
+		subRing := ring.ShuffleShard(uid, shardSize)
+		sr := subRing.(*Ring)
+
+		// find some instance in subring
+		var instanceID string
+		for id := range sr.ringDesc.Ingesters {
```

Comment:
[nit] I would run this test for every instance in the ring. It's an easy change to do, but would increase our confidence.

Reply:
Great idea.

Reply:
Sure enough, it found bugs!

Reply:
It turned out that our ring tests can be flaky, because our tests could produce conflicting tokens. After fixing token generation in tests, test is now reliable.